docs/man/man3/AdcRead.3
.TH "AdcRead" 3 "Tue May 17 2022" "EPS MCU" \" -*- nroff -*-
.ad l
.nh
.SH NAME
AdcRead \- Læser og gemmer ADC målinger\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <AdcRead\&.h>\fP
.SS "Classes"

.in +1c
.ti -1c
.RI "struct \fBadc_reading\fP"
.br
.ti -1c
.RI "struct \fBadc_thread_args\fP"
.br
.in -1c
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBAdcName\fP { \fBBAT_V_1\fP, \fBBAT_I_1\fP, \fBMPPT_V_1\fP, \fBMPPT_I_1\fP, \fBSUB_V_1\fP, \fBSUB_I_1\fP }"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBAdcRead\fP ()=default"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static \fBAdcRead::adc_reading\fP \fBDoRead\fP ()"
.br
.RI "Kalder AsyncADC med et antal ADC'er på baggrund af nuværende state\&. Nye måling bliver lagt i CircularBuffer'en adc_buffer_\&. "
.ti -1c
.RI "static void \fBReadU16\fP (\fBadc_thread_args\fP *args)"
.br
.RI "Måler et givent antal gange på et givent device fra \fBadc_thread_args\fP og skriver middelværdi til \fBadc_thread_args\fP dest\&. "
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static CircularBuffer< \fBadc_reading\fP, \fBADC_BUFFER_SIZE\fP > \fBadc_buffer_\fP"
.br
.in -1c
.SS "Static Private Member Functions"

.in +1c
.ti -1c
.RI "static AnalogIn * \fBGetDevice\fP (\fBAdcName\fP name)"
.br
.RI "Omdanner AdcName enum til pointer til ny AnalogIn object for respektivt ADC ben\&. "
.ti -1c
.RI "static float \fBReadFloat\fP (AnalogIn *device)"
.br
.ti -1c
.RI "static void \fBAsyncADC\fP (int adc_count)"
.br
.RI "Starter et givent antal threads der hver kører \fBReadU16()\fP med dets givet ADC ben\&. "
.in -1c
.SS "Static Private Attributes"

.in +1c
.ti -1c
.RI "static \fBAdcName\fP \fBadc_list_\fP [6]"
.br
.in -1c
.SH "Detailed Description"
.PP 
Læser og gemmer ADC målinger\&. 
.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBAdcRead::AdcName\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIBAT_V_1 \fP\fP
.TP
\fB\fIBAT_I_1 \fP\fP
.TP
\fB\fIMPPT_V_1 \fP\fP
.TP
\fB\fIMPPT_I_1 \fP\fP
.TP
\fB\fISUB_V_1 \fP\fP
.TP
\fB\fISUB_I_1 \fP\fP
.SH "Constructor & Destructor Documentation"
.PP 
.SS "AdcRead::AdcRead ()\fC [default]\fP"

.SH "Member Function Documentation"
.PP 
.SS "void AdcRead::AsyncADC (int adc_count)\fC [static]\fP, \fC [private]\fP"

.PP
Starter et givent antal threads der hver kører \fBReadU16()\fP med dets givet ADC ben\&. 
.PP
\fBParameters\fP
.RS 4
\fIadc_count\fP Antal ADC'er at måle på 
.RE
.PP

.SS "\fBAdcRead::adc_reading\fP AdcRead::DoRead ()\fC [static]\fP"

.PP
Kalder AsyncADC med et antal ADC'er på baggrund af nuværende state\&. Nye måling bliver lagt i CircularBuffer'en adc_buffer_\&. 
.PP
\fBReturns\fP
.RS 4
\fBAdcRead::adc_reading\fP Nye måling 
.RE
.PP

.SS "AnalogIn * AdcRead::GetDevice (\fBAdcRead::AdcName\fP name)\fC [static]\fP, \fC [private]\fP"

.PP
Omdanner AdcName enum til pointer til ny AnalogIn object for respektivt ADC ben\&. 
.PP
\fBParameters\fP
.RS 4
\fIname\fP enum for det ADC ben der skal måles på 
.RE
.PP
\fBReturns\fP
.RS 4
AnalogIn* pointer til oprettet device 
.RE
.PP

.SS "static float AdcRead::ReadFloat (AnalogIn * device)\fC [static]\fP, \fC [private]\fP"

.SS "void AdcRead::ReadU16 (\fBadc_thread_args\fP * args)\fC [static]\fP"

.PP
Måler et givent antal gange på et givent device fra \fBadc_thread_args\fP og skriver middelværdi til \fBadc_thread_args\fP dest\&. 
.PP
\fBParameters\fP
.RS 4
\fIargs\fP \fBadc_thread_args\fP med AnalogIn pointer til devicet og uint16_t pointer til resultatet 
.RE
.PP

.SH "Member Data Documentation"
.PP 
.SS "CircularBuffer< \fBAdcRead::adc_reading\fP, \fBADC_BUFFER_SIZE\fP > AdcRead::adc_buffer_\fC [static]\fP"

.SS "\fBAdcName\fP AdcRead::adc_list_[6]\fC [static]\fP, \fC [private]\fP"


.SH "Author"
.PP 
Generated automatically by Doxygen for EPS MCU from the source code\&.


lib/AdcRead/AdcRead.cpp
//
// Created by EIT-416 on 01/04/2022.
//

#include "AdcRead.h"
#include "../Config/ProjectConfig.h"
#include <mbed.h>
#include <rtos.h>
#include "Fsm.h"

CircularBuffer<AdcRead::adc_reading, ADC_BUFFER_SIZE> AdcRead::adc_buffer_;
uint16_t adc_read_list[ADC_NUM_CHANNELS];

Thread *t_adc_read[ADC_NUM_CHANNELS];
AdcRead::adc_thread_args arg_list[ADC_NUM_CHANNELS];

/**
 * @brief Kalder AsyncADC med et antal ADC'er på baggrund af nuværende state.
 *        Nye måling bliver lagt i CircularBuffer'en adc_buffer_
 * 
 * @return AdcRead::adc_reading Nye måling
 */
AdcRead::adc_reading AdcRead::DoRead() {
  Fsm::State current_state = Fsm::GetCurrentState();
  adc_reading new_read{};

  if (current_state > Fsm::LOW) {
    int adc_count = ADC_NUM_CHANNELS;
    AsyncADC(adc_count);
    new_read = {
        adc_read_list[0],
        adc_read_list[1],
        adc_read_list[2],
        adc_read_list[3],
        adc_read_list[4],
        adc_read_list[5],
        time(NULL)
    };
  } else {
    int adc_count = 2;
    AsyncADC(adc_count);
    new_read = {
        adc_read_list[0],
        adc_read_list[1],
        0,
        0,
        0,
        0,
        time(NULL)
    };
  }

  adc_buffer_.push(new_read);
  return new_read;
}

/**
 * @brief Starter et givent antal threads der hver kører ReadU16()
 *        med dets givet ADC ben
 * 
 * @param adc_count Antal ADC'er at måle på
 */
void AdcRead::AsyncADC(int adc_count) {
  for (int i = 0; i < adc_count; ++i) {
    arg_list[i] = {
        GetDevice((AdcName) i),
        &adc_read_list[i]
    };
    t_adc_read[i] = new Thread();
    t_adc_read[i]->start(callback(ReadU16, &arg_list[i]));
  }

  for (int i = 0; i < adc_count; ++i) {
    t_adc_read[i]->join();
  }

  for (int i = 0; i < adc_count; ++i) {
    delete t_adc_read[i];
  }
}

/**
 * @brief Omdanner AdcName enum til pointer til ny AnalogIn object
 *        for respektivt ADC ben
 * 
 * @param name enum for det ADC ben der skal måles på
 * @return AnalogIn* pointer til oprettet device
 */
AnalogIn *AdcRead::GetDevice(AdcRead::AdcName name) {

  AnalogIn *device_ptr;

  switch (name) {
    case BAT_V_1:device_ptr = new AnalogIn(BAT_V_1_PIN);
      break;

    case BAT_I_1:device_ptr = new AnalogIn(BAT_I_1_PIN);
      break;

    case MPPT_V_1:device_ptr = new AnalogIn(MPPT_V_1_PIN);
      break;

    case MPPT_I_1:device_ptr = new AnalogIn(MPPT_I_1_PIN);
      break;

    case SUB_V_1:device_ptr = new AnalogIn(SUB_V_1_PIN);
      break;

    case SUB_I_1:device_ptr = new AnalogIn(SUB_I_1_PIN);
      break;
  }

  return device_ptr;
}

/**
 * @brief Måler et givent antal gange på et givent device 
 *        fra adc_thread_args og skriver middelværdi til 
 *        adc_thread_args dest
 * 
 * @param args adc_thread_args med AnalogIn pointer til devicet og 
 *             uint16_t pointer til resultatet
 */
void AdcRead::ReadU16(adc_thread_args *args) {
  float sum = 0;
  for (int i = 0; i < ADC_MEASURE_LOOPS; i++) {
//    adc_mutex.lock();
    sum += (float) args->device->read_u16() / (float) ADC_MEASURE_LOOPS;
//    adc_mutex.unlock();
  }

  delete args->device;

  if (sum == 0) {
    sum = 1;
  }

  *args->dest = (uint16_t) sum;
}
